File: tests/leetcode/x/elixir/0005.ex
defmodule Main do
  def expand(s, left, right) do
    if left >= 0 and right < byte_size(s) and :binary.at(s, left) == :binary.at(s, right) do
      expand(s, left - 1, right + 1)
    else
      {left + 1, right - left - 1}
    end
  end

  def longest_palindrome(s) do
    n = byte_size(s)
    {best_start, best_len} =
      Enum.reduce(0..max(n - 1, -1), {0, if(n > 0, do: 1, else: 0)}, fn i, {best_start, best_len} ->
        {start1, len1} = expand(s, i, i)
        {best_start, best_len} = if len1 > best_len, do: {start1, len1}, else: {best_start, best_len}
        {start2, len2} = expand(s, i, i + 1)
        if len2 > best_len, do: {start2, len2}, else: {best_start, best_len}
      end)

    binary_part(s, best_start, best_len)
  end

  def main do
    lines = IO.read(:stdio, :eof) |> String.split(~r/\r?\n/, trim: false)
    if lines != [] do
      t = hd(lines) |> String.trim() |> String.to_integer()
      ss = Enum.take(tl(lines) ++ List.duplicate("", t), t)
      IO.write(Enum.map_join(ss, "\n", &longest_palindrome/1))
    end
  end
end

Main.main()

File: tests/leetcode/x/lua/0005.lua
local function expand(s, left, right)
  while left >= 1 and right <= #s and s:sub(left, left) == s:sub(right, right) do
    left = left - 1
    right = right + 1
  end
  return left + 1, right - left - 1
end

local function longestPalindrome(s)
  local bestStart = 1
  local bestLen = #s > 0 and 1 or 0
  for i = 1, #s do
    local start1, len1 = expand(s, i, i)
    if len1 > bestLen then
      bestStart = start1
      bestLen = len1
    end
    local start2, len2 = expand(s, i, i + 1)
    if len2 > bestLen then
      bestStart = start2
      bestLen = len2
    end
  end
  return s:sub(bestStart, bestStart + bestLen - 1)
end

local lines = {}
for line in io.lines() do
  table.insert(lines, line)
end
if #lines == 0 then
  return
end
local t = tonumber(lines[1])
local out = {}
for i = 1, t do
  out[#out + 1] = longestPalindrome(lines[i + 1] or '')
end
io.write(table.concat(out, '\n'))
